lib/address_geocoder.rb

```diff
@@ -1,2 +1,168 @@
+require 'yaml'
+
 class AddressGeocoder
-end
\ No newline at end of file
+  attr_accessor :api_key, :country, :state, :city, :postal_code, :street
+  attr_reader :countries
+  attr_writer :google_response, :former_address, :refined_address
+
+  def initialize(opt = {})
+    # 1. Initialize variables
+    @api_key         = opt[:api_key]
+    @country         = opt[:country]
+    @state           = opt[:state]
+    @city            = opt[:city]
+    @postal_code     = opt[:postal_code]
+    @street          = opt[:street]
+    @countries       = YAML.load_file('lib/address_geocoder/countries.yaml')['countries']['country']
+    @refined_address = {
+      :country     => nil,
+      :state       => nil,
+      :city        => nil,
+      :postal_code => nil,
+      :street      => nil
+    }
+  end
+
+  def validate_address
+    return refine_address
+  end
+
+  private
+  def refine_address
+    if situation_has_changed
+      # 1. Initialize refined address
+      @refined_address[:country] = match_country
+
+      # 2. Loop through the levels (once one works break the loop)
+      get_format_levels.each do |level_of_search|
+
+        # 2.1 Make call to Google
+        attempts = 0
```

Comment:
really good to retry to hit remote API.
It may be an other private method to make refine_address smaller